lib/sync/ephy-open-tabs-manager.c

```diff
@@ -0,0 +1,257 @@
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with Epiphany.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "config.h"
+#include "ephy-open-tabs-manager.h"
+
+#include "ephy-embed-container.h"
+#include "ephy-embed-shell.h"
+#include "ephy-settings.h"
+#include "ephy-synchronizable-manager.h"
+
+struct _EphyOpenTabsManager {
+  GObject parent_instance;
+
+  /* A list of EphyOpenTabsRecord objects describing the open tabs
+   * of other sync clients. This is updated at every sync. */
+  GSList *remote_records;
+};
+
+static void ephy_synchronizable_manager_iface_init (EphySynchronizableManagerInterface *iface);
+
+G_DEFINE_TYPE_WITH_CODE (EphyOpenTabsManager, ephy_open_tabs_manager, G_TYPE_OBJECT,
+                         G_IMPLEMENT_INTERFACE (EPHY_TYPE_SYNCHRONIZABLE_MANAGER,
+                                                ephy_synchronizable_manager_iface_init))
+
+static void
+ephy_open_tabs_manager_dispose (GObject *object)
+{
+  EphyOpenTabsManager *self = EPHY_OPEN_TABS_MANAGER (object);
+
+  g_slist_free_full (self->remote_records, g_object_unref);
+  self->remote_records = NULL;
+
+  G_OBJECT_CLASS (ephy_open_tabs_manager_parent_class)->dispose (object);
+}
+
+static void
+ephy_open_tabs_manager_class_init (EphyOpenTabsManagerClass *klass)
+{
+  GObjectClass *object_class = G_OBJECT_CLASS (klass);
+
+  object_class->dispose = ephy_open_tabs_manager_dispose;
+}
+
+static void
+ephy_open_tabs_manager_init (EphyOpenTabsManager *self)
+{
+}
+
+EphyOpenTabsManager *
+ephy_open_tabs_manager_new (void)
+{
+  return EPHY_OPEN_TABS_MANAGER (g_object_new (EPHY_TYPE_OPEN_TABS_MANAGER, NULL));
+}
+
+EphyOpenTabsRecord *
+ephy_open_tabs_manager_get_local_tabs (EphyOpenTabsManager *self)
+{
+  EphyOpenTabsRecord *local_tabs;
+  WebKitFaviconDatabase *database;
+  EphyEmbedShell *embed_shell;
+  GList *windows;
+  GList *tabs;
+  char *favicon;
+  char *client_id;
+  char *client_name;
+  const char *title;
+  const char *url;
+
+  g_return_val_if_fail (EPHY_IS_OPEN_TABS_MANAGER (self), NULL);
+
+  embed_shell = ephy_embed_shell_get_default ();
+  windows = gtk_application_get_windows (GTK_APPLICATION (embed_shell));
+  database = webkit_web_context_get_favicon_database (ephy_embed_shell_get_web_context (embed_shell));
+  client_id = g_settings_get_string (EPHY_SETTINGS_SYNC, EPHY_PREFS_SYNC_CLIENT_ID);
+  client_name = g_strdup_printf ("%s on Epiphany", client_id);
+  local_tabs = ephy_open_tabs_record_new (client_id, client_name);
+
+  for (GList *l = windows; l && l->data; l = l->next) {
+    tabs = ephy_embed_container_get_children (l->data);
+
+    for (GList *t = tabs; t && t->data; t = t->next) {
+      title = ephy_embed_get_title (t->data);
+
+      if (!g_strcmp0 (title, "Blank page") || !g_strcmp0 (title, "Most Visited"))
```

Comment:
Hm, this only works for English, right? You probably need to pass it through gettext _().

Even so, that's fragile. I know the layering here is awkward, but is there someplace that we could define constants for these? Then it can be used like this:

`#define BLANK_PAGE_TITLE N_("Blank page")`

`!g_strcmp0 (title, _(BLANK_PAGE_TITLE))`